mllib/src/main/scala/org/apache/spark/ml/tree/impl/RandomForest.scala

```diff
@@ -1113,4 +1114,77 @@ private[ml] object RandomForest extends Logging {
     }
   }
 
+  /**
+   * Given a Random Forest model, compute the importance of each feature.
+   * This generalizes the idea of "Gini" importance to other losses,
+   * following the explanation of Gini importance from "Random Forests" documentation
+   * by Leo Breiman and Adele Cutler, and following the implementation from scikit-learn.
+   *
+   * This feature importance is calculated as follows:
+   *  - Average over trees:
+   *     - importance(feature j) = sum (over nodes which split on feature j) of the gain,
+   *       where gain is scaled by the number of instances passing through node
+   *     - Normalize importances for tree based on total number of training instances used
+   *       to build tree.
+   *  - Normalize feature importance vector to sum to 1.
+   *
+   * Note: This should not be used with Gradient-Boosted Trees.  It only makes sense for
+   *       independently trained trees.
+   * Note: This is returned as a Map since models do not store the number of features.
+   *       That should be corrected in the future.
+   * @param trees  Unweighted forest of trees
+   * @return  Feature importance values.  Returned as map from feature index to importance.
+   */
+  private[ml] def featureImportances(trees: Array[DecisionTreeModel]): Map[Int, Double] = {
+    val totalImportances = new OpenHashMap[Int, Double]()
+    trees.foreach { tree =>
+      // Aggregate feature importance vector for this tree
+      val importances = new OpenHashMap[Int, Double]()
+      computeFeatureImportance(tree.rootNode, importances)
+      // Normalize importance vector for this tree, and add it to total.
+      val treeNorm = tree.rootNode.impurityStats.count
```

Comment:
In sklearn decision trees, you can either (a) normalize the feature importance vector by # instances, or (b) normalize the importance vector to sum to 1.  When that is called to compute feature importance for forests, sklearn does (b) by default.  Originally, I had implemented this using (a), but I switched to (b) to match sklearn.